simpl/incubator/rest.py

```diff
@@ -190,3 +154,41 @@ def wrapped(*args, **kwargs):
                 bottle.abort(400, str(exc))
         return wrapped
     return deco
+
+
+def body(schema=None, types=None, required=False, default=None):
+    """Decorator to parse and validate API body.
+
+    :keyword schema: callable that accepts raw data and returns the coerced (or
+        unchanged) data if it is valid. It should raise an error if the data is
+        not valid.
+    :keyword types: supported content types (default is ['application/json'])
+    :keyword required: if true and no body specified will raise an error.
+    :keyword default: default value to return if no body supplied
+
+    Note: only json types are supported.
+    """
+    if not types:
+        types = ['application/json']
+    if not all('json' in t for t in types):
+        raise NotImplementedError("Only 'json' body supported.")
+
+    def wrap(fxn):
+        """Return a decorated callable."""
+        def wrapped(*args, **kwargs):
+            """Callable to called when the decorated function is called."""
+            body = bottle.request.json
+            if required and not body:
+                bottle.abort(400, "Call body cannot be empty")
+            if body is None:
+                body = default
+            if schema:
+                try:
+                    body = schema(body)
+                except (KeyboardInterrupt, SystemExit):
+                    raise  # don't catch and ignore attempts to end the app
+                except Exception as exc:
+                    bottle.abort(400, str(exc))
+            return fxn(*args, body=body, **kwargs)
```

Comment:
I understand "less invasive" as being less restrictive on the decorated callable's signature. I agree that adding a kwarg (as opposed to adding an arg) is easier to do, easier to move around or handle using `**kwargs`, and can be done with more tolerance for argument ordering (to some degree, since kwargs can still receive values sent in as positional argument). But I do think it does also have the downside that it can conflict with kwargs or args of the same name if they are already in the decorated callable signature causing `multiple values for keyword argument 'body'` errors. And while that can be partially resolved by checking for duplicate kwargs like so:
```python
def decorator(fxn):
    def wrapper(*args, **kwargs):
        kwargs.setdefault('body', new_data)  # or `kwargs['body'] = new_data
        return fxn(*args, **kwargs)
    return wrapper
```
there are cons to either of those fixes that could cause hard-to-find "_where is the data I am passing in the `body` argument getting changed?!_" type of bugs.       

If this was a "generic" decorator that we wanted to be provide for others to use on any function, then we would probably try very hard to not modify the signature of the caller. But since this is only going to be used on a callable that is most likely designed to work with `bottle` and `simpl` and that `simpl` is "opinionated" in how it should be used, then I feel that adding a first argument to the callable is more **explicit** and therefore less complex and less prone to errors.

```python
@rest.body(schema=WIDGET)
def post_widget(body):
    data = save(body)
    return data
```

In following with trying to be more explicit, it might also make sense to leave a `body` argument unchanged if the call is not coming from a bottle route:
```python
def decorator(fxn):
    def wrapper(*args, **kwargs):
        # Only override body if we're in a bottle call
        if in_bottle_route_with_body():
            body = bottle.request.json
        else:
            body = args.pop(0)
        return fxn(body, *args, **kwargs)
    return wrapper
```
This way the callable can still be called directly from other code or in tests.